### ISSUE TITLE
Can't open include file

### ISSUE THREAD

**Issue by thomtsib** [M-oster/Advanced-Ultimate-Box-Maker#1]
Hello,

I tried to use the instuctions but I have these messages as warnings:

WARNING: Can't open include file 'BOSL2/common.scad'. 
WARNING: Can't open include file 'BOSL2/arrays.scad'. 

WARNING: Ignoring unknown function 'select' in file Advanced-Ultimate-Box-Maker-main.scad, line 329 

Do you know how to find these files? Also do you know why the openscad can't find select function?

**Comment by M-oster** [comment]
Hi Thomtsib,
at first thanks for your interest in my program. When you  have followed the instructions step by step, the prog will work.

to your question: "select" is a function of the module "arrays" in the BOSL2-lib.

I think you have done alle the steps of the instructions, at first check the lib-folder of Openscad. 
Please open Openscad and check  under "File", "show library path" (or similar, I have the german version)
there you must see the both directories of the used libs.
![1-libdir](https://user-images.githubusercontent.com/29811839/139957979-39324feb-a8f6-492a-a6f0-42e219df3252.png)


OpenSCAD uses three library /locations/, the installation library, built-in library, and user-defined libraries.
 1. The /Installation/ library location is the |libraries| directory under the directory where OpenSCAD is installed.
 2. The /Built-In/ library location is O/S dependent. Since version 2014.03, it can be opened in the system specific file manager using the "File->Show Library Folder..." menu entry.
      * Windows: |My Documents\OpenSCAD\libraries|
      * Linux: |$HOME/.local/share/OpenSCAD/libraries|
      * Mac OS X: |$HOME/Documents/OpenSCAD/libraries|
 3. The /User-Defined/ library path can be created using the OPENSCADPATH Environment Variable to point to the library(s).
    OPENSCADPATH can contain multiple directories in case you have  library collections in more than one place, separate directories  with a semi-colon for Windows, and a colon for Linux/Mac OS. For example:
        Windows:
        |C:\Users\Username\Documents\OpenSCAD\MyLib;C:\ThingiverseStuff\OpenSCAD Things;D:\test_stuff|

         Linux/Mac OS: |/usr/lib:/home/mylib:.|

      OpenSCAD must be restarted to recognize any change to the OPENSCADPATH Environment Variable.

Its the easiest way to choose the 2nd point and put the libs in the directory "My Documents\OpenSCAD\libraries".

I try to help you by describing the installation process more detailed: (my descr. is for windows, for linux I can't do support.)
1. download of the recent openscad-prog from https://files.openscad.org/OpenSCAD-2021.01-x86-64-Installer.exe
2. install the prog with standard parameters.
3. download the both required libs, the BOSL2-lib from https://github.com/revarbat/BOSL2/  and the NopSCADlib from 
https://github.com/nophead/NopSCADlib. both as zip-file
4. and the last one,  download my box-maker-prog as zip from github.
5.create a directory named "libraries" under "My Documents\OpenSCAD\libraries
6. Then unzip all the files (incl. the subdirectories of cause) of the box-maker-prog in a project directory for OpenScad (..My 
Documentsusers\documents\openscad\Box-maker) for example
7. unzip the both library-files into the created directory "My Documents\OpenSCAD\libraries"

Thats it..
in summary it lookes like this:

![2-openscad-dir](https://user-images.githubusercontent.com/29811839/139958095-debf840c-9352-4635-a408-62713037e595.png)

for a check I have downloaded the github-files und unzipped them into the created folder "Advanced-Ultimate-Box-Maker-github". But this name doesn't matter.
The library files are in subdir "libraries".

![3-libs](https://user-images.githubusercontent.com/29811839/139958339-c73dbb9e-09ee-4f74-9534-211b95bdaa5c.png)


and the content of the  box-maker directory:

![4-prog-dir](https://user-images.githubusercontent.com/29811839/139958390-405298fe-9dab-4dab-9e4c-5510c7078763.png)

Thats the unzipped content of the zip-file from github. Then the program must work.

I hope I could help you.

**Comment by thomtsib** [comment]
Hello,

my pleasure to use your program, I am excited from your work. However, I followed all these steps and I had a double check and all are same as yours. I think the problem is that the array.scad and common.scad are missing from the ...Documents\OpenSCAD\libraries\BOSL2  path as you can see at the image.
![BOSL2](https://user-images.githubusercontent.com/10002016/140020925-da476535-4940-487e-a14c-b5ee0c4b1add.png)
Do you have these files on your pc ?

**Comment by M-oster** [comment]
ah, ok. I have found the issue. My fault was that I have not downloaded the most recent BOSL2-lib for my test. 
revarbat, the programmer of the lib, has uploaded a new version on 1st of Nov. and they don't have these two scad-files any longer.
I will check it if the used functions are in another file of this lib. I have included only the files which are needed. If the whole lib will be included the openscad-program will be rather slow.
Please be patient while, I will search for a solution.
the BOSL2 lib is in beta, ok, but the BOSL1 lib does not have the array functions I need.

**Comment by M-oster** [comment]
Hi Thom, we will go the normal way and include the BOSL2-lib as the programmer suggests. He has rearranged the modules so that some functions moved to other scad-files. But with the general include all is working well.
Please delete in the box-maker-scad-files alle these lines including the BOSL2-lib:
include <BOSL2/common.scad>
include <BOSL2/vectors.scad>
include <BOSL2/arrays.scad>
include <BOSL2/math.scad>
include <BOSL2/strings.scad>

and replace it by the general include:
**include <BOSL2/std.scad>**

you must change that for the "Advanced-Ultimate-Box-Maker" project and for the "Mini-PowerSupply" Project.

Sorry for my late understanding but I didn't expect that the lib has changed in its structure.
During the day I will update the fixed code here in github and in thingiverse

**Comment by thomtsib** [comment]
Seem to be ok!!!
